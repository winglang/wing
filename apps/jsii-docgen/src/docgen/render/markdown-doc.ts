import { Language } from "../..";
import { DocsSchema } from "../schema";

/**
 * Options for defining a markdown header.
 */
export interface MarkdownHeaderOptions {
  /**
   * Title to be displayed.
   */
  readonly title?: string;

  /**
   * Superscript.
   *
   * @default - No superscript
   */
  readonly sup?: string;

  /**
   * Preformat the header.
   *
   * @default false
   */
  readonly pre?: boolean;

  /**
   * Strikethough the title.
   *
   * @default false
   */
  readonly strike?: boolean;
}

/**
 * Options for defining a markdown element.
 */
export interface MarkdownOptions {
  /**
   * Markdown header.
   *
   * @default - No header.
   */
  readonly header?: MarkdownHeaderOptions;

  /**
   * Id of the element.
   *
   * @default - The title will be used.
   */
  readonly id?: string;
}

/**
 * Markdown element.
 */
export class MarkdownDocument {
  /**
   * An empty markdown element.
   */
  public static readonly EMPTY = new MarkdownDocument();

  /**
   * Sanitize markdown reserved characters from external input.
   */
  public static sanitize(line: string): string {
    let sanitized = line.trim();

    if (line.startsWith("-")) {
      sanitized = sanitized.substring(1, line.length).trim();
    }

    // JSDoc tags that use "@" must be escaped with backticks to
    // avoid getting parsed as another JSDoc tag. Sanitize away
    // these backticks to avoid double escaping.
    if (line.startsWith("`") && line.endsWith("`")) {
      sanitized = sanitized.substring(1, line.length - 1).trim();

      // unescape < and > inside code lines
      sanitized = sanitized.replace(/&lt;/g, "<").replace(/&gt;/g, ">");
    }

    return sanitized;
  }

  /**
   * Remove newlines from markdown.
   */
  public static removeNewlines(line: string): string {
    return line.replace(/\n/g, " ");
  }

  public static bold(text: string): string {
    return `**${text}**`;
  }

  public static pre(text: string): string {
    // using <code> instead of backticks since this allows links
    return `<code>${text}</code>`;
  }

  public static italic(text: string) {
    return `*${text}*`;
  }

  private readonly _lines = new Array<string>();
  private readonly _sections = new Array<MarkdownDocument>();

  private readonly id?: string;
  private readonly header?: string;

  constructor(private readonly options: MarkdownOptions = {}) {
    this.id = options.id ?? options.header?.title;
    this.header = this.formatHeader();
  }

  /**
   * Render a docs element into the markdown.
   */
  public docs(docs: DocsSchema, language?: Language) {
    if (docs.inflight && docs.inflight !== "true") {
      const sanitized = MarkdownDocument.sanitize(docs.inflight);
      this.lines(
        MarkdownDocument.bold("Inflight client:") +
          " [" +
          sanitized +
          "](#" +
          sanitized +
          ")"
      );
      this.lines("");
    }
    if (docs.summary) {
      this.lines(MarkdownDocument.sanitize(docs.summary));
      this.lines("");
    }
    if (docs.remarks) {
      this.lines(MarkdownDocument.sanitize(docs.remarks));
      this.lines("");
    }

    if (docs.links) {
      for (const link of docs.links) {
        this.quote(`[${link}](${link})`);
      }
    }

    if (docs.example) {
      if (!language) {
        throw new Error(
          "language must be provided if docs.example has been specified"
        );
      }
      const example = new MarkdownDocument({
        id: `${this.options.id}.example`,
      });
      example.lines(MarkdownDocument.italic("Example"), "");
      example.code(language.toString(), docs.example);
      example.lines("");
      this.section(example);
    }
  }

  public table(data: string[][]) {
    const numColumns = data[0].length;
    const header = data[0];
    const rows = data.slice(1);
    this.lines("| " + header.map(this.escapePipes).join(" | ") + " |");
    this.lines("|" + " --- |".repeat(numColumns));
    for (const row of rows) {
      this.lines("| " + row.map(this.escapePipes).join(" | ") + " |");
    }
    this.lines("");
  }

  public quote(line: string) {
    this.lines(`> ${line}`);
    this.lines("");
  }

  public title(line: string, headerSize: number) {
    this.lines(`${"#".repeat(headerSize)} ${line}`);
    this.lines("");
  }

  public bullet(line: string) {
    this.lines(`- ${line}`);
  }

  public code(language: string, ...snippet: string[]) {
    this.lines(
      `\`\`\`${language}`,
      ...snippet.map((s) => s.replace(/&lt;/g, "<").replace(/&gt;/g, ">")),
      "```"
    );
    this.lines("");
  }

  public lines(...lines: string[]) {
    this._lines.push(...lines);
  }

  public split() {
    this.lines("---");
    this.lines("");
  }

  public section(section: MarkdownDocument) {
    this._sections.push(section);
  }

  public render(headerSize: number = 0): string {
    const content: string[] = [];

    if (this.header) {
      if (headerSize > 6) {
        // headers are mapped to `h1-h6` html elements.
        // passed that, markdown just renders `#` signs.
        // lets see if and when we'll hit this limit.
        throw new Error("Unable to render markdown. Header limit (6) reached.");
      }

      const heading = `${"#".repeat(headerSize)} ${this.header}`;

      // temporary hack to avoid breaking Construct Hub
      const headerSpan = !!process.env.HEADER_SPAN;
      if (headerSpan) {
        content.push(
          `${heading} <span data-heading-title="${this.options.header?.title}" data-heading-id="${this.id}"></span>`
        );
      } else {
        content.push(
          `${heading} <a name="${this.options.header?.title}" id="${this.id}"></a>`
        );
      }
      content.push("");
    }

    for (const line of this._lines) {
      content.push(`${line}`);
    }

    for (const section of this._sections) {
      content.push(section.render(headerSize + 1));
    }
    return content.join("\n");
  }

  private formatHeader(): string | undefined {
    if (!this.options.header?.title) {
      return undefined;
    }
    let caption = this.options.header.title;

    if (this.options.header?.pre ?? false) {
      caption = `\`${caption}\``;
    }

    if (this.options.header?.strike ?? false) {
      caption = `~~${caption}~~`;
    }

    if (this.options.header?.sup) {
      caption = `${caption}<sup>${this.options.header?.sup}</sup>`;
    }

    return caption;
  }

  private escapePipes(line: string): string {
    return line.replace(/\|/g, "\\|");
  }
}
