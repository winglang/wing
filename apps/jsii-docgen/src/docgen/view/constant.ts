import * as reflect from "jsii-reflect";
import { PropertySchema } from "../schema";
import { Transpile } from "../transpile/transpile";
import { Property } from "./property";

export class Constant {
  private readonly constant: Property;
  constructor(transpile: Transpile, property: reflect.Property) {
    this.constant = new Property(transpile, property);
  }

  public toJson(): PropertySchema {
    return {
      ...this.constant.toJson(),
      const: true,
    };
  }
}
