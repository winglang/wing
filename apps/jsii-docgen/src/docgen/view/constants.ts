import * as reflect from "jsii-reflect";
import { PropertySchema } from "../schema";
import { Transpile } from "../transpile/transpile";
import { Constant } from "./constant";

export class Constants {
  private readonly constants: Constant[];
  constructor(transpile: Transpile, properties: reflect.Property[]) {
    this.constants = properties
      .filter((p) => !p.protected && p.const)
      .map((p) => new Constant(transpile, p));
  }

  public toJson(): PropertySchema[] {
    return this.constants.map((c) => c.toJson());
  }
}
