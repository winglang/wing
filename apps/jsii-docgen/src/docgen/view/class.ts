import * as reflect from "jsii-reflect";
import { Constants } from "./constants";
import { Initializer } from "./initializer";
import { InstanceMethods } from "./instance-methods";
import { Properties } from "./properties";
import { StaticFunctions } from "./static-functions";
import { HIDDEN_METHODS, HIDDEN_STATIC } from "./wing-filters";
import { ClassSchema, extractDocs } from "../schema";
import {
  Transpile,
  TranspiledClass,
  TranspiledType,
} from "../transpile/transpile";

const CONSTRUCT_CLASS = "constructs.Construct";

export class Class {
  public static isConstruct(klass: reflect.ClassType): boolean {
    if (klass.fqn === CONSTRUCT_CLASS) return true;

    if (!klass.base) {
      return false;
    }

    return this.isConstruct(klass.base);
  }

  private readonly transpiled: TranspiledClass;

  private readonly initializer?: Initializer;
  private readonly instanceMethods: InstanceMethods;
  private readonly staticFunctions: StaticFunctions;
  private readonly constants: Constants;
  private readonly properties: Properties;
  private readonly interfaces: TranspiledType[];

  constructor(
    private readonly transpile: Transpile,
    private readonly klass: reflect.ClassType
  ) {
    if (klass.initializer) {
      this.initializer = new Initializer(transpile, klass.initializer);
    }
    this.instanceMethods = new InstanceMethods(
      transpile,
      klass.allMethods.filter((m) => !HIDDEN_METHODS.includes(m.name))
    );
    this.staticFunctions = new StaticFunctions(
      transpile,
      klass.allMethods.filter((m) => !HIDDEN_STATIC.includes(m.name))
    );
    this.constants = new Constants(transpile, klass.allProperties);
    this.properties = new Properties(transpile, klass.allProperties);
    this.interfaces = klass.interfaces.map((iface) =>
      this.transpile.type(iface)
    );
    this.transpiled = transpile.class(klass);
  }

  public toJson(): ClassSchema {
    const name = this.transpiled.type.fqn.split(".").pop()!;
    const namespace = this.transpiled.type.namespace;
    return {
      initializer: this.initializer?.toJson(),
      interfaces: this.interfaces.map((iface) => iface.toJson()),
      instanceMethods: this.instanceMethods.toJson(),
      staticMethods: this.staticFunctions.toJson(),
      constants: this.constants.toJson(),
      properties: this.properties.toJson(),
      fqn: this.transpiled.type.fqn,
      displayName:
        name === "Util" && namespace
          ? // capitalized namespace
            namespace.charAt(0).toUpperCase() + namespace.slice(1)
          : name,
      id: this.klass.fqn,
      docs: extractDocs(this.klass.docs),
    };
  }
}
