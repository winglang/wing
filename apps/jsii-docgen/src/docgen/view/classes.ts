import * as reflect from "jsii-reflect";
import { ClassSchema } from "../schema";
import { Transpile } from "../transpile/transpile";
import { Class } from "./class";

export class Classes {
  private readonly classes: Class[];
  constructor(transpile: Transpile, classes: reflect.ClassType[]) {
    this.classes = classes
      .filter((c) => !Class.isConstruct(c))
      .map((c) => new Class(transpile, c));
  }

  public toJson(): ClassSchema[] {
    return this.classes.map((klass) => klass.toJson());
  }
}
