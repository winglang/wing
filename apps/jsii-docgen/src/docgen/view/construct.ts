import * as reflect from "jsii-reflect";
import { Class } from "./class";
import { Interface } from "./interface";
import { ConstructSchema } from "../schema";
import { Transpile } from "../transpile/transpile";

export class Construct {
  private readonly construct: Class;
  private readonly inflightInterface?: Interface;
  constructor(
    transpile: Transpile,
    klass: reflect.ClassType,
    inflight?: Interface
  ) {
    this.inflightInterface = inflight;
    this.construct = new Class(transpile, klass);
  }

  public toJson(): ConstructSchema {
    return {
      ...this.construct.toJson(),
      inflight: this.inflightInterface?.toJson(),
    };
  }
}
