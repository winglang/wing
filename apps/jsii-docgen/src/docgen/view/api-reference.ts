import * as reflect from "jsii-reflect";
import { Classes } from "./classes";
import { Constructs, WingClassType } from "./constructs";
import { Enums } from "./enums";
import { Interfaces } from "./interfaces";
import { Structs } from "./structs";
import { VISIBLE_SUBMODULES } from "./wing-filters";
import { MarkdownDocument } from "../render/markdown-doc";
import { ApiReferenceSchema, getInflight, isSkipped } from "../schema";
import { Transpile } from "../transpile/transpile";

/**
 * Render an API reference based on the jsii assembly.
 */
export class ApiReference {
  private readonly constructs: Constructs;
  private readonly structs: Structs;
  private readonly interfaces: Interfaces;
  private readonly classes: Classes;
  private readonly enums: Enums;
  constructor(
    transpile: Transpile,
    assembly: reflect.Assembly,
    submodule?: reflect.Submodule,
    allSubmodules?: boolean
  ) {
    let classes: WingClassType[];
    let interfaces: reflect.InterfaceType[];
    let enums: reflect.EnumType[];

    let submodules = assembly.submodules;

    // TODO WING SDK HACK
    const wingSdkHacks = assembly.name === "@winglang/sdk";
    if (wingSdkHacks) {
      submodules = submodules.filter((s) =>
        VISIBLE_SUBMODULES.includes(s.name)
      );
    }
    if (allSubmodules ?? false) {
      classes = this.sortByName([
        ...assembly.classes,
        ...flatMap(submodules, (submod) => [...submod.classes]),
      ]);
      interfaces = this.sortByName([
        ...assembly.interfaces,
        ...flatMap(submodules, (submod) => [...submod.interfaces]),
      ]);
      enums = this.sortByName([
        ...assembly.enums,
        ...flatMap(submodules, (submod) => [...submod.enums]),
      ]);
    } else {
      classes = this.sortByName(
        submodule ? submodule.classes : assembly.classes
      );
      interfaces = this.sortByName(
        submodule ? submodule.interfaces : assembly.interfaces
      );
      enums = this.sortByName(submodule ? submodule.enums : assembly.enums);
    }

    classes = classes.filter((item) => !isSkipped(item.docs));
    interfaces = interfaces.filter((item) => !isSkipped(item.docs));
    enums = enums.filter((item) => !isSkipped(item.docs));

    classes.forEach((c) => {
      const inflight = getInflight(c.docs);
      if (inflight) {
        c.inflightFqn = MarkdownDocument.sanitize(inflight);
      }
    });

    const inflightFqns: string[] = classes
      .map(({ inflightFqn }) => inflightFqn)
      .filter((fqn) => typeof fqn === "string") as string[];

    this.constructs = new Constructs(
      transpile,
      classes,
      this.searchableInterfaces([...assembly.allInterfaces])
    );
    this.classes = new Classes(transpile, classes);
    this.structs = new Structs(transpile, interfaces);
    this.interfaces = new Interfaces(
      transpile,
      this.filterOutInflightInterfaces(interfaces, inflightFqns)
    );
    this.enums = new Enums(transpile, enums);
  }

  /**
   * Generate JSON.
   */
  public toJson(): ApiReferenceSchema {
    return {
      constructs: this.constructs.toJson(),
      classes: this.classes.toJson(),
      structs: this.structs.toJson(),
      interfaces: this.interfaces.toJson(),
      enums: this.enums.toJson(),
    };
  }

  private searchableInterfaces(
    interfaces: reflect.InterfaceType[]
  ): Record<string, reflect.InterfaceType> {
    return interfaces.reduce(
      (acc, iface) => ({ ...acc, [iface.fqn.replace(/ /g, "-")]: iface }),
      {}
    );
  }

  private filterOutInflightInterfaces(
    interfaces: reflect.InterfaceType[],
    inflightFqns: string[]
  ): reflect.InterfaceType[] {
    return interfaces.filter(
      (iface) => !inflightFqns.includes(iface.fqn.replace(/ /g, "-"))
    );
  }

  private sortByName<Type extends reflect.Type>(arr: readonly Type[]): Type[] {
    return [...arr].sort((s1, s2) => s1.name.localeCompare(s2.name));
  }
}

function flatMap<T, U>(
  xs: readonly T[],
  fn: (value: T, index: number, array: readonly T[]) => U[]
): U[] {
  return Array.prototype.concat(...xs.map(fn));
}
