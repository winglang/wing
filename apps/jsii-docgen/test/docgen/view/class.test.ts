import * as reflect from "jsii-reflect";
import { Language } from "../../../src";
import { MarkdownRenderer } from "../../../src/docgen/render/markdown-render";
import { Class } from "../../../src/docgen/view/class";
import { LANGUAGE_SPECIFIC } from "../../../src/docgen/view/documentation";
import { Assemblies } from "../assemblies";

const assembly: reflect.Assembly = Assemblies.instance.withoutSubmodules;

const metadata = {
  packageName: assembly.name,
  packageVersion: assembly.version,
};

const findClass = (): reflect.ClassType => {
  if (assembly.classes[0]) {
    return assembly.classes[0];
  }
  throw new Error("Assembly does not contain a class");
};

test.each(Language.values().map((l) => l.name))("%s snapshot", (language) => {
  const { transpile } = LANGUAGE_SPECIFIC[language.toString()];
  const markdown = new MarkdownRenderer({ language, ...metadata });
  const klass = new Class(transpile, findClass()).toJson();
  expect(klass).toMatchSnapshot();
  expect(markdown.visitClass(klass).render()).toMatchSnapshot();
});
