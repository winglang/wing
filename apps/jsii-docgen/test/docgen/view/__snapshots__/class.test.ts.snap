// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csharp snapshot 1`] = `
Object {
  "constants": Array [],
  "displayName": "AuthorizationToken",
  "docs": Object {
    "links": Array [
      "https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html",
    ],
    "summary": "Authorization token to access private ECR repositories in the current environment via Docker CLI.",
  },
  "fqn": "Amazon.CDK.AWS.ECR.AuthorizationToken",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken",
  "initializer": undefined,
  "instanceMethods": Array [],
  "interfaces": Array [],
  "properties": Array [],
  "staticMethods": Array [
    Object {
      "displayName": "GrantRead",
      "docs": Object {
        "summary": "Grant access to retrieve an authorization token.",
      },
      "fqn": "Amazon.CDK.AWS.ECR.AuthorizationToken.GrantRead",
      "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead",
      "parameters": Array [
        Object {
          "default": undefined,
          "displayName": "Grantee",
          "docs": Object {},
          "fqn": "Amazon.CDK.AWS.ECR.AuthorizationToken.GrantRead.parameter.Grantee",
          "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
          "optional": undefined,
          "type": Object {
            "formattingPattern": "%",
            "types": Array [
              Object {
                "displayName": "IGrantable",
                "fqn": "Amazon.CDK.AWS.IAM.IGrantable",
                "id": "@aws-cdk/aws-iam.IGrantable",
                "packageName": "@aws-cdk/aws-iam",
                "packageVersion": "99.99.99",
              },
            ],
          },
        },
      ],
      "usage": "using Amazon.CDK.AWS.ECR;

AuthorizationToken.GrantRead(IGrantable Grantee);",
    },
  ],
}
`;

exports[`csharp snapshot 2`] = `
" AuthorizationToken <a name=\\"AuthorizationToken\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken\\"></a>

Authorization token to access private ECR repositories in the current environment via Docker CLI.

> [https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html](https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html)


# Static Methods <a name=\\"Static Methods\\" id=\\"Static Methods\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>GrantRead</code> | Grant access to retrieve an authorization token. |

---

**<code>GrantRead</code>**
\`\`\`csharp
using Amazon.CDK.AWS.ECR;

AuthorizationToken.GrantRead(IGrantable Grantee);
\`\`\`

Grant access to retrieve an authorization token.

| **Parameter** | **Type** | **Description** |
| --- | --- | --- |
| <code>Grantee</code> | <code>Amazon.CDK.AWS.IAM.IGrantable</code> | *No description.* |


"
`;

exports[`go snapshot 1`] = `
Object {
  "constants": Array [],
  "displayName": "AuthorizationToken",
  "docs": Object {
    "links": Array [
      "https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html",
    ],
    "summary": "Authorization token to access private ECR repositories in the current environment via Docker CLI.",
  },
  "fqn": "github.com/aws/aws-cdk-go/awscdkawsecr.AuthorizationToken",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken",
  "initializer": undefined,
  "instanceMethods": Array [],
  "interfaces": Array [],
  "properties": Array [],
  "staticMethods": Array [
    Object {
      "displayName": "GrantRead",
      "docs": Object {
        "summary": "Grant access to retrieve an authorization token.",
      },
      "fqn": "github.com/aws/aws-cdk-go/awscdkawsecr.AuthorizationToken.GrantRead",
      "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead",
      "parameters": Array [
        Object {
          "default": undefined,
          "displayName": "grantee",
          "docs": Object {},
          "fqn": "github.com/aws/aws-cdk-go/awscdkawsecr.AuthorizationToken.GrantRead.parameter.grantee",
          "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
          "optional": undefined,
          "type": Object {
            "formattingPattern": "%",
            "types": Array [
              Object {
                "displayName": "IGrantable",
                "fqn": "github.com/aws/aws-cdk-go/awscdk/awsiam/awscdkawsiam.IGrantable",
                "id": "@aws-cdk/aws-iam.IGrantable",
                "packageName": "@aws-cdk/aws-iam",
                "packageVersion": "99.99.99",
              },
            ],
          },
        },
      ],
      "usage": "import \\"github.com/aws/aws-cdk-go/awscdkawsecr\\"

awscdkawsecr.AuthorizationToken_GrantRead(grantee IGrantable)",
    },
  ],
}
`;

exports[`go snapshot 2`] = `
" AuthorizationToken <a name=\\"AuthorizationToken\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken\\"></a>

Authorization token to access private ECR repositories in the current environment via Docker CLI.

> [https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html](https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html)


# Static Methods <a name=\\"Static Methods\\" id=\\"Static Methods\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>GrantRead</code> | Grant access to retrieve an authorization token. |

---

**<code>GrantRead</code>**
\`\`\`go
import \\"github.com/aws/aws-cdk-go/awscdkawsecr\\"

awscdkawsecr.AuthorizationToken_GrantRead(grantee IGrantable)
\`\`\`

Grant access to retrieve an authorization token.

| **Parameter** | **Type** | **Description** |
| --- | --- | --- |
| <code>grantee</code> | <code>github.com/aws/aws-cdk-go/awscdk/awsiam/awscdkawsiam.IGrantable</code> | *No description.* |


"
`;

exports[`java snapshot 1`] = `
Object {
  "constants": Array [],
  "displayName": "AuthorizationToken",
  "docs": Object {
    "links": Array [
      "https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html",
    ],
    "summary": "Authorization token to access private ECR repositories in the current environment via Docker CLI.",
  },
  "fqn": "software.amazon.awscdk.services.ecr.AuthorizationToken",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken",
  "initializer": undefined,
  "instanceMethods": Array [],
  "interfaces": Array [],
  "properties": Array [],
  "staticMethods": Array [
    Object {
      "displayName": "grantRead",
      "docs": Object {
        "summary": "Grant access to retrieve an authorization token.",
      },
      "fqn": "software.amazon.awscdk.services.ecr.AuthorizationToken.grantRead",
      "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead",
      "parameters": Array [
        Object {
          "default": undefined,
          "displayName": "grantee",
          "docs": Object {},
          "fqn": "software.amazon.awscdk.services.ecr.AuthorizationToken.grantRead.parameter.grantee",
          "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
          "optional": undefined,
          "type": Object {
            "formattingPattern": "%",
            "types": Array [
              Object {
                "displayName": "IGrantable",
                "fqn": "software.amazon.awscdk.services.iam.IGrantable",
                "id": "@aws-cdk/aws-iam.IGrantable",
                "packageName": "@aws-cdk/aws-iam",
                "packageVersion": "99.99.99",
              },
            ],
          },
        },
      ],
      "usage": "import software.amazon.awscdk.services.ecr.AuthorizationToken;

AuthorizationToken.grantRead(IGrantable grantee)",
    },
  ],
}
`;

exports[`java snapshot 2`] = `
" AuthorizationToken <a name=\\"AuthorizationToken\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken\\"></a>

Authorization token to access private ECR repositories in the current environment via Docker CLI.

> [https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html](https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html)


# Static Methods <a name=\\"Static Methods\\" id=\\"Static Methods\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>grantRead</code> | Grant access to retrieve an authorization token. |

---

**<code>grantRead</code>**
\`\`\`java
import software.amazon.awscdk.services.ecr.AuthorizationToken;

AuthorizationToken.grantRead(IGrantable grantee)
\`\`\`

Grant access to retrieve an authorization token.

| **Parameter** | **Type** | **Description** |
| --- | --- | --- |
| <code>grantee</code> | <code>software.amazon.awscdk.services.iam.IGrantable</code> | *No description.* |


"
`;

exports[`python snapshot 1`] = `
Object {
  "constants": Array [],
  "displayName": "AuthorizationToken",
  "docs": Object {
    "links": Array [
      "https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html",
    ],
    "summary": "Authorization token to access private ECR repositories in the current environment via Docker CLI.",
  },
  "fqn": "aws_cdk.aws_ecr.AuthorizationToken",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken",
  "initializer": undefined,
  "instanceMethods": Array [],
  "interfaces": Array [],
  "properties": Array [],
  "staticMethods": Array [
    Object {
      "displayName": "grant_read",
      "docs": Object {
        "summary": "Grant access to retrieve an authorization token.",
      },
      "fqn": "aws_cdk.aws_ecr.AuthorizationToken.grant_read",
      "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead",
      "parameters": Array [
        Object {
          "default": undefined,
          "displayName": "grantee",
          "docs": Object {},
          "fqn": "aws_cdk.aws_ecr.AuthorizationToken.grant_read.parameter.grantee",
          "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
          "optional": undefined,
          "type": Object {
            "formattingPattern": "%",
            "types": Array [
              Object {
                "displayName": "IGrantable",
                "fqn": "aws_cdk.aws_iam.IGrantable",
                "id": "@aws-cdk/aws-iam.IGrantable",
                "packageName": "@aws-cdk/aws-iam",
                "packageVersion": "99.99.99",
              },
            ],
          },
        },
      ],
      "usage": "import aws_cdk.aws_ecr

aws_cdk.aws_ecr.AuthorizationToken.grant_read(
  grantee: IGrantable
)",
    },
  ],
}
`;

exports[`python snapshot 2`] = `
" AuthorizationToken <a name=\\"AuthorizationToken\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken\\"></a>

Authorization token to access private ECR repositories in the current environment via Docker CLI.

> [https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html](https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html)


# Static Methods <a name=\\"Static Methods\\" id=\\"Static Methods\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>grant_read</code> | Grant access to retrieve an authorization token. |

---

**<code>grant_read</code>**
\`\`\`python
import aws_cdk.aws_ecr

aws_cdk.aws_ecr.AuthorizationToken.grant_read(
  grantee: IGrantable
)
\`\`\`

Grant access to retrieve an authorization token.

| **Parameter** | **Type** | **Description** |
| --- | --- | --- |
| <code>grantee</code> | <code>aws_cdk.aws_iam.IGrantable</code> | *No description.* |


"
`;

exports[`typescript snapshot 1`] = `
Object {
  "constants": Array [],
  "displayName": "AuthorizationToken",
  "docs": Object {
    "links": Array [
      "https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html",
    ],
    "summary": "Authorization token to access private ECR repositories in the current environment via Docker CLI.",
  },
  "fqn": "@aws-cdk/aws-ecr.AuthorizationToken",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken",
  "initializer": undefined,
  "instanceMethods": Array [],
  "interfaces": Array [],
  "properties": Array [],
  "staticMethods": Array [
    Object {
      "displayName": "grantRead",
      "docs": Object {
        "summary": "Grant access to retrieve an authorization token.",
      },
      "fqn": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead",
      "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead",
      "parameters": Array [
        Object {
          "default": undefined,
          "displayName": "grantee",
          "docs": Object {},
          "fqn": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
          "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
          "optional": undefined,
          "type": Object {
            "formattingPattern": "%",
            "types": Array [
              Object {
                "displayName": "IGrantable",
                "fqn": "@aws-cdk/aws-iam.IGrantable",
                "id": "@aws-cdk/aws-iam.IGrantable",
                "packageName": "@aws-cdk/aws-iam",
                "packageVersion": "99.99.99",
              },
            ],
          },
        },
      ],
      "usage": "import { AuthorizationToken } from '@aws-cdk/aws-ecr'

AuthorizationToken.grantRead(grantee: IGrantable)",
    },
  ],
}
`;

exports[`typescript snapshot 2`] = `
" AuthorizationToken <a name=\\"AuthorizationToken\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken\\"></a>

Authorization token to access private ECR repositories in the current environment via Docker CLI.

> [https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html](https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html)


# Static Methods <a name=\\"Static Methods\\" id=\\"Static Methods\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>grantRead</code> | Grant access to retrieve an authorization token. |

---

**<code>grantRead</code>**
\`\`\`typescript
import { AuthorizationToken } from '@aws-cdk/aws-ecr'

AuthorizationToken.grantRead(grantee: IGrantable)
\`\`\`

Grant access to retrieve an authorization token.

| **Parameter** | **Type** | **Description** |
| --- | --- | --- |
| <code>grantee</code> | <code>@aws-cdk/aws-iam.IGrantable</code> | *No description.* |


"
`;

exports[`wing snapshot 1`] = `
Object {
  "constants": Array [],
  "displayName": "AuthorizationToken",
  "docs": Object {
    "links": Array [
      "https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html",
    ],
    "summary": "Authorization token to access private ECR repositories in the current environment via Docker CLI.",
  },
  "fqn": "@aws-cdk/aws-ecr.AuthorizationToken",
  "id": "@aws-cdk/aws-ecr.AuthorizationToken",
  "initializer": undefined,
  "instanceMethods": Array [],
  "interfaces": Array [],
  "properties": Array [],
  "staticMethods": Array [
    Object {
      "displayName": "grantRead",
      "docs": Object {
        "summary": "Grant access to retrieve an authorization token.",
      },
      "fqn": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead",
      "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead",
      "parameters": Array [
        Object {
          "default": undefined,
          "displayName": "grantee",
          "docs": Object {},
          "fqn": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
          "id": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
          "optional": undefined,
          "type": Object {
            "formattingPattern": "%",
            "types": Array [
              Object {
                "displayName": "IGrantable",
                "fqn": "@aws-cdk/aws-iam.IGrantable",
                "id": "@aws-cdk/aws-iam.IGrantable",
                "packageName": "@aws-cdk/aws-iam",
                "packageVersion": "99.99.99",
              },
            ],
          },
        },
      ],
      "usage": "bring { AuthorizationToken } from \\"@aws-cdk/aws-ecr\\"

AuthorizationToken.grantRead(grantee: IGrantable)",
    },
  ],
}
`;

exports[`wing snapshot 2`] = `
" AuthorizationToken <a name=\\"AuthorizationToken\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken\\"></a>

Authorization token to access private ECR repositories in the current environment via Docker CLI.

> [https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html](https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html)


# Static Methods <a name=\\"Static Methods\\" id=\\"Static Methods\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code>grantRead</code> | Grant access to retrieve an authorization token. |

---

**<code>grantRead</code>**
\`\`\`wing
bring { AuthorizationToken } from \\"@aws-cdk/aws-ecr\\"

AuthorizationToken.grantRead(grantee: IGrantable)
\`\`\`

Grant access to retrieve an authorization token.

| **Parameter** | **Type** | **Description** |
| --- | --- | --- |
| <code>grantee</code> | <code>@aws-cdk/aws-iam.IGrantable</code> | *No description.* |


"
`;
